mm2src/coins/utxo/utxo_common.rs

```diff
@@ -58,126 +61,266 @@ lazy_static! {
+            .await?
+            .into_iter(),
+    );
+    account_balance.addresses.extend(
+        check_balance_of_hd_account_unknown_addresses(coin, hd_wallet, hd_account, external_addresses)
+            .await?
+            .into_iter(),
+    );
 
-    fn ticker(&self) -> &str { self.ticker }
+    // Request balances of the internal addresses.
+    account_balance.addresses.extend(
+        balance_of_hd_account_used_addresses(coin, hd_account, internal_addresses)
+            .await?
+            .into_iter(),
+    );
+    account_balance.addresses.extend(
+        check_balance_of_hd_account_unknown_addresses(coin, hd_wallet, hd_account, internal_addresses)
+            .await?
+            .into_iter(),
+    );
 
-    fn priv_key(&self) -> PrivKeyBuildPolicy<'_> { self.priv_key.clone() }
+    Ok(account_balance)
 }
 
-pub async fn utxo_arc_from_conf_and_params<T>(
-    ctx: &MmArc,
-    ticker: &str,
-    conf: &Json,
-    activation_params: UtxoActivationParams,
-    priv_key: PrivKeyBuildPolicy<'_>,
-    constructor: impl Fn(UtxoArc) -> T + Send + 'static,
-) -> UtxoCoinBuildResult<T>
+/// Requests balances of HD account addresses that have been used already.
+pub async fn balance_of_hd_account_used_addresses<T>(
+    coin: &T,
+    hd_account: &UtxoHDAccount,
+    change: bool,
+) -> BalanceResult<Vec<HDAddressBalance>>
 where
-    T: AsRef<UtxoCoinFields> + UtxoCommonOps + Send + Sync + 'static,
+    T: UtxoCommonOps
+        + UtxoHDWalletOps
+        + HDWalletBalanceOps<HDWallet = UtxoHDWallet, HDAccount = UtxoHDAccount, Address = Address>
+        + Sync,
 {
-    let builder = UtxoArcBuilder::new(ctx, ticker, conf, activation_params.clone(), priv_key);
-    let utxo_arc = builder.build().await?;
-    let coin = constructor(utxo_arc.clone());
-
-    if let Some(merge_params) = activation_params.utxo_merge_params {
-        let weak = utxo_arc.downgrade();
-        let merge_loop = merge_utxo_loop(
-            weak,
-            merge_params.merge_at,
-            merge_params.check_every,
-            merge_params.max_merge_at_once,
-            constructor,
-        );
-        info!("Starting UTXO merge loop for coin {}", ticker);
-        spawn(merge_loop);
+    let non_empty_addresses_number = if change {
+        hd_account.internal_addresses_number
+    } else {
+        hd_account.external_addresses_number
+    };
+    let mut balances = Vec::with_capacity(non_empty_addresses_number as usize);
+
+    for address_id in 0..non_empty_addresses_number {
+        if address_id >= ChildNumber::HARDENED_FLAG {
+            warn!(
+                "address_id '{}' is too large. Stop the account '{}' loop",
+                address_id, hd_account.account_id
+            );
+            break;
+        }
+
+        let HDAddress {
+            address,
+            derivation_path,
+        } = coin.derive_address(hd_account, address_id, change)?;
+
+        let balance = coin.known_address_balance(&address).await?;
```

Comment:
Requesting balances sequentially can become a problem if users will have many addresses after a long time. It's preferable to start using batch requests here.

Reply:
I splitted the `wallet_balance` into `hd_account_balance` and `check_hd_account_balance`, added pagging options to `hd_account_balance`. Could you please re-review?

Reply:
I meant using JSON RPC batch requests https://www.jsonrpc.org/specification#batch. E.g. request all balances of known addresses from Electrum in a single batch. 
Please check if using batches will be more effective - compare the time required to perform e.g. 10 balance requests sequentially or in the batch. If the difference is significant, consider refactoring to batch requests in the next iteration.